pynbody/snapshot/gadgethdf.py

```diff
@@ -365,9 +365,9 @@ def _get_hdf_dataset(self, particle_group, hdf_name):
     def _get_cosmo_factors(hdf, arr_name) :
         """Return the cosmological factors for a given array"""
         match = [s for s in GadgetHDFSnap._get_hdf_allarray_keys(hdf) if ((s.endswith("/"+arr_name)) & ('PartType' in s))]
-        if arr_name == 'Mass' and len(match) == 0:
-            # mass stored in header. We're out in the cold on our own.
-            return units.Unit('1.0'), units.h**-1
+        if len(match) == 0:
+                # no match
+            return False, False
```

Comment:
This could break other code which relies on `_get_cosmo_factors` (it is called from several places). I think it would be better to revert this but issue a warning from here to match the type of warning you are issuing below.

Reply:
Since the mass dataset could be named both "Mass" and "Masses", this "if" statement was triggered in both the following cases:
1- If the mass dataset is missing and masses are stored in the header (as, I assume, it was originally conceived)
2- If the mass dataset is present but has a different name.

As a consequence, in __init_unit_information(), if one wants to check for the presence, e.g., of both "Mass" and "Masses" a factor h^-2 is accumulated if the case where "no match is found" is not handled.
I opted to modify _get_cosmo_factors() in order to disentangle these two cases and also because I did not find any other usage of it in the whole folder.

Reply:
I'm not fully following, but I agree the current logic is rather tortured. My point is different -- I don't think it ever makes sense to return a tuple of `False, False`. This could cause huge confusion to callers. If the condition should never be encountered, it should raise an exception. If the condition can lead to a default guess, that default should be returned and a warning issued.

You are right it isn't called from elsewhere, I misread that -- but still, to enable its potential reuse and save caller confusion, I think the function should either return units, or raise an exception, but never return a boolean.